KIPs/kip-82.md

```diff
@@ -0,0 +1,334 @@
+    MinimumStake: int = 5000000               # "reward.minimumstake" parameter (in KLAY)
+    DeferredTxFee: bool = true    # "reward.deferredtxfee" parameter
+    UseKIP82: bool = true         # "reward.usekip82" parameter (new)
+
+    # "reward.ratio" parameter (e.g. "34/54/12")
+    CnRatio: int = 34
+    KgfRatio: int = 54
+    KirRatio: int = 12
+    TotalRatio: int = CnRatio + KgfRatio + KirRatio
+
+    # "reward.kip82ratio" parameter (e.g. "20/80") (new)
+    CnMintedBasicRatio: int = 20
+    CnMintedStakeRatio: int = 80
+    CnMintedTotalRatio: int = CnMintedBasicRatio + CnMintedStakeRatio
+
+# CN staking status and KGF/KIR addresses.
+# Corresponds to Klaytn's reward.StakingInfo struct.
+class StakingInfo:
+    KIRAddr: string = ""
+    KGFAddr: string = ""
+    Nodes: List[ConsolidatedNode] = []
+
+# Staking information merged under the same CN.
+# Sometimes a node would register multiple NodeAddrs
+# in which each entry has different StakingAddr and the same RewardAddr.
+# We treat those entries with common RewardAddr as one node.
+# Corresponds to Klaytn's reward.ConsolidatedNode struct.
+class ConsolidatedNode:
+    NodeAddrs: List[string] = []
+    StakingAddrs: List[string] = []
+    RewardAddr: string = "" # The common reward address of the CN
+    StakingAmount: int = 0  # Total staking amount across StakingAddrs (in KLAY)
+
+# Reward disribution details.
+class RewardSpec:
+    Minted: int = 0   # The amount minted
+    Fee: int = 0      # Total tx fee spent
+    Burnt: int = 0    # The amount burnt
+    Proposer: int = 0    # The amount allocated to the block proposer
+    Stakers: int = 0  # Total amount allocated to stakers
+    Kgf: int = 0      # The amount allocated to KGF
+    Kir: int = 0      # The amount allocated to KIR
+    Rewards: Dict[string, int] = {}  # Mapping from reward recipient to amounts
+
+# Distributes a given block's reward to CNs and KGF/KIR funds.
+# corresponds to Klaytn's reward.RewardDistributor.DistributeBlockReward().
+#
+# - state is the StateDB to apply the rewards.
+# - header is a Header instance.
+# - config is a RewardConfig instance.
+# - staking_info is a StakingInfo instance.
+def distribute_block_reward(state, header, config, staking_info):
+    spec = calc_deferred_reward(header, config, staking_info)
+
+    for addr, amount in spec.Rewrads:
+        state.AddBalance(addr, amount)
+
+# Calculates the deferred rewards, which are determined at the end of block processing.
+# Used in reward distribution.
+# Returns a RewardSpec.
+def calc_deferred_reward(header, config, staking_info):
+    minted = config.MintingAmount
+    fee, burnt = calc_fee_resource(header, config)
+
+    proposer, stakers, kgf, kir, split_rem = split_reward(header, config, minted, fee)
+    shares, share_rem = calc_stake_shares(config, staking_info, stakers)
+
+    kgf += split_rem
+    kgf += share_rem
+
+    spec = RewardSpec()
+    spec.Minted = minted
+    spec.Fee = fee
```

Comment:
@blukat29 
spec.Fee is `Total tx fee spent` in RewardSpec, but it seems `calc_state_shares()` returns remaining fees.
Which is right?

Reply:
Fixed `calc_fee_resource` to return `(total_fee, reward_fee, burnt_fee)`.